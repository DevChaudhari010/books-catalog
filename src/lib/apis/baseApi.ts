import axios from "axios";

const host = process.env.API_SERVER;

const axiosInstance = axios.create({
  baseURL: host + "/api/v1",
});

export default axiosInstance;
