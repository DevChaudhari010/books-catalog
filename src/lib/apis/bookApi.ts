import { IBook } from "@types";
import baseApi from "./baseApi";

export type GetBooksResponse = {
  books: IBook[] | null;
  error: boolean;
};

const getBooks: () => Promise<GetBooksResponse> = async () => {
  try {
    const response = await baseApi({ method: "get", url: "books" });
    return { books: response.data, error: false };
  } catch (error) {
    return { error: true, books: null };
  }
};

export type GetBookDetailsResponse = {
  book: IBook | null;
  error: boolean;
};

const getBookDetails: (id: string) => Promise<GetBookDetailsResponse> = async (
  bookId
) => {
  try {
    const response = await baseApi({
      method: "get",
      url: `books/${bookId}`,
    });
    return { book: response.data, error: false };
  } catch (error) {
    return { error: true, book: null };
  }
};

export default { getBooks, getBookDetails };
