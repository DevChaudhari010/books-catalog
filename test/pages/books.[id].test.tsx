import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Book, { getStaticProps, getStaticPaths } from "@pages/books/[id]";
import { IBook } from "@types";
import { bookApi } from "@lib/apis";

jest.mock("@lib/apis");

const books: IBook[] = [
  { _id: "1", title: "Test 1" },
  { _id: "2", title: "Test 2" },
];

const book: IBook = {
  _id: "1",
  title: "Test 1",
  description: "Test Desc",
  year: 1990,
};

describe("Book Details page", () => {
  test("matches snapshot for book Details", () => {
    const { asFragment } = render(<Book book={book} error={false} />, {});
    expect(asFragment()).toMatchSnapshot();
  });
  test("matches snapshot for error", () => {
    const { asFragment } = render(<Book book={null} error={true} />, {});
    expect(asFragment()).toMatchSnapshot();
  });
  test("Check book details exists", () => {
    const { getByText } = render(<Book book={book} error={false} />, {});
    expect(getByText("Test 1")).toBeInTheDocument();
    expect(getByText("Test Desc")).toBeInTheDocument();
    expect(getByText("1990")).toBeInTheDocument();
  });
  test("Check error messgae exists on Error", () => {
    const { getByText } = render(<Book book={null} error={true} />, {});
    expect(
      getByText("Failed to load book Details. Please try again after sometime")
    ).toBeInTheDocument();
  });
  test("Check getStaticProps", async () => {
    const response = { book, error: false };
    //@ts-ignore
    bookApi.getBookDetails.mockImplementation(() => Promise.resolve(response));
    const staticProps = await getStaticProps({});
    expect(bookApi.getBookDetails).toBeCalled();
    expect(staticProps).toEqual(
      expect.objectContaining({
        props: response,
      })
    );
  });
  test("Check getStaticProps no book found", async () => {
    const response = { book: null, error: false };
    //@ts-ignore
    bookApi.getBookDetails.mockImplementation(() => Promise.resolve(response));
    const staticProps = await getStaticProps({});
    expect(bookApi.getBookDetails).toBeCalled();
    expect(staticProps).toEqual(
      expect.objectContaining({
        notFound: true,
      })
    );
  });
  test("Check getStaticPaths", async () => {
    const response = { books, error: false };
    //@ts-ignore
    bookApi.getBooks.mockImplementation(() => Promise.resolve(response));
    const staticPaths = await getStaticPaths({});
    expect(bookApi.getBooks).toBeCalled();
    expect(staticPaths).toEqual(
      expect.objectContaining({
        paths: response.books.map(({ _id }) => ({
          params: { id: _id.toString() },
        })),
      })
    );
  });
});
