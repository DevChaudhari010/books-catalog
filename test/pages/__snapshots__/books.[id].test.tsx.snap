// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book Details page matches snapshot for book Details 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <h2>
      Test 1
    </h2>
    <span>
      1990
    </span>
    <p>
      Test Desc
    </p>
  </div>
</DocumentFragment>
`;

exports[`Book Details page matches snapshot for error 1`] = `
<DocumentFragment>
  <div>
    Failed to load book Details. Please try again after sometime
  </div>
</DocumentFragment>
`;
